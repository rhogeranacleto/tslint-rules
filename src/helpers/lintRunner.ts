import { Configuration, Linter, LintResult } from 'tslint';

interface IMaria {
	src: string;
	// tslint:disable-next-line:no-any
	rule: any;
	fileName?: string;
	fix?: boolean;
}

export function lintRunner(data: IMaria) {

	const { src, rule, fileName = `/tmp/test-${Date.now()}.snap`, fix = false } = data;

	const linter = new Linter({ fix });

	linter.lint(
		fileName,
		src,
		Configuration.parseConfigFile({
			rules: {
				[rule.name || rule]: [true]
			},
			rulesDirectory: 'src'
		})
	);

	return linter.getResult();
}

export function getReplacements(result: LintResult) {

	if (result.fixes) {

		return result.fixes.map(fix => {

			const fixJson = fix.toJson().fix;

			if (fixJson) {

				if (Array.isArray(fixJson)) {

					return fixJson.map(replace => replace.innerText);
				}

				return [fixJson.innerText];
			}
		});
	}
}