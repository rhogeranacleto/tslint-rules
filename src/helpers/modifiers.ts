export function removeLastEmptyLineOfTheBlock(block: string) {

	const lines = block.split('\n');
	const bracket = lines.pop();

	for (let i = lines.length - 1; i > 0; i--) {

		if (lines[i].trim()) {

			break;
		}

		lines.pop();
	}

	lines.push(bracket as string);

	return lines.join('\n');
}