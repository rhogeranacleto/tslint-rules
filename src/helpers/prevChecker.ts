// tslint:disable-next-line:no-implicit-dependencies
import { getPreviousStatement, getPreviousToken } from 'tsutils';
import ts from 'typescript';

export function prevStatementChecker(node: ts.Statement, sourceFile: ts.SourceFile) {

	const prev = getPreviousStatement(node);

	if (prev) {

		return checkPrevLine(node, sourceFile, prev);
	}

	return false;
}

export function prevTokenChecker(node: ts.Node, sourceFile: ts.SourceFile) {

	const prev = getPreviousToken(node);

	if (prev) {

		return checkPrevLine(node, sourceFile, prev);
	}

	return false;
}

function checkPrevLine(node: ts.Node, sourceFile: ts.SourceFile, prev: ts.Node) {

	const startLine = ts.getLineAndCharacterOfPosition(sourceFile, node.getStart()).line;
	const prevLine = ts.getLineAndCharacterOfPosition(sourceFile, prev.getEnd()).line;

	return prevLine === (startLine - 1);
}