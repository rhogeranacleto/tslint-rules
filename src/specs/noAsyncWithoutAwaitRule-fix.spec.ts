import { getFileText, lintRunner } from '../helpers/lintRunner';

const rule = 'no-async-without-await';

describe('noAsyncWithoutAwait fix', () => {

	describe('Functions', () => {

		it(`should fail when the async function doesn't have an await`, () => {

			const src = `
			async function a(){
				let b = 1;
				console.log(b);
			}`;

			const result = lintRunner({ src, rule, fix: true });

			expect(result.errorCount).toBe(0);

			expect(getFileText(result)).toMatchSnapshot();
		});

		it('should fail when the async function has an await inside a arrow function declaration', () => {

			const src = `
			async function a(){
				let b = 1;
				let b = async () => {
					await fetch();
				};
			}`;
			const result = lintRunner({ src, rule, fix: true });

			expect(result.errorCount).toBe(0);

			expect(getFileText(result)).toMatchSnapshot();
		});

		it('should fail when the async function has an await inside a regular function declaration', () => {

			const src = `
			async function a(){
				let b = 1;
				async function f() {
					await fetch();
				};
			}`;
			const result = lintRunner({ src, rule, fix: true });

			expect(result.errorCount).toBe(0);

			expect(getFileText(result)).toMatchSnapshot();
		});

		it('should fail when the function is not at the top level', () => {

			const src = `
			function a(){
				let b = 1;
				async function f() {
					fetch();
				};
			}`;

			const result = lintRunner({ src, rule, fix: true });

			expect(result.errorCount).toBe(0);

			expect(getFileText(result)).toMatchSnapshot();
		});
	});

	describe('Arrow functions', () => {

		it(`should fail when the async arrow function doesn't have an await`, () => {

			const src = `
			const a = async () => {
				let b = 1;
				console.log(b);
			}`;

			const result = lintRunner({ src, rule, fix: true });

			expect(result.errorCount).toBe(0);

			expect(getFileText(result)).toMatchSnapshot();
		});
	});

	describe('Classes', () => {

		it('should fail when async class method has no await', () => {

			const src = `
			class A {
				async b() {
					console.log(1);
				}
			}
		`;
			const result = lintRunner({ src, rule, fix: true });

			expect(result.errorCount).toBe(0);

			expect(getFileText(result)).toMatchSnapshot();
		});

		it('should fail when async class method has no await', () => {

			const src = `
			class A {
				public async b() {
					console.log(1);
				}
			}
		`;
			const result = lintRunner({ src, rule, fix: true });

			expect(result.errorCount).toBe(0);

			expect(getFileText(result)).toMatchSnapshot();
		});

		it('should check classes that are inside functions', () => {

			const src = `
			async () => {
				await a();
				class A {
					async b() {
						console.log(1);
					}
				}
			};
		`;
			const result = lintRunner({ src, rule, fix: true });

			expect(result.errorCount).toBe(0);

			expect(getFileText(result)).toMatchSnapshot();
		});
	});

	describe('return statements', () => {

		it('should fail when the async function has a return statement', () => {

			const src = `
			async function a() {
				let b = 1;
				let a = () => {
				return 1;
				}
			}
		`;
			const result = lintRunner({ src, rule, fix: true });

			expect(result.errorCount).toBe(0);

			expect(getFileText(result)).toMatchSnapshot();
		});
	});
});