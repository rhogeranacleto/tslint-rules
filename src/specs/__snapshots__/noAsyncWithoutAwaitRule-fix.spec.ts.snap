// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`noAsyncWithoutAwait fix Arrow functions should fail when the async arrow function doesn't have an await 1`] = `
"
			const a = () => {
				let b = 1;
				console.log(b);
			}"
`;

exports[`noAsyncWithoutAwait fix Classes should check classes that are inside functions 1`] = `
"
			async () => {
				await a();
				class A {
					b() {
						console.log(1);
					}
				}
			};
		"
`;

exports[`noAsyncWithoutAwait fix Classes should fail when async class method has no await 1`] = `
"
			class A {
				b() {
					console.log(1);
				}
			}
		"
`;

exports[`noAsyncWithoutAwait fix Classes should fail when async class method has no await 2`] = `
"
			class A {
				public b() {
					console.log(1);
				}
			}
		"
`;

exports[`noAsyncWithoutAwait fix Functions should fail when the async function doesn't have an await 1`] = `
"
			function a(){
				let b = 1;
				console.log(b);
			}"
`;

exports[`noAsyncWithoutAwait fix Functions should fail when the async function has an await inside a arrow function declaration 1`] = `
"
			function a(){
				let b = 1;
				let b = async () => {
					await fetch();
				};
			}"
`;

exports[`noAsyncWithoutAwait fix Functions should fail when the async function has an await inside a regular function declaration 1`] = `
"
			function a(){
				let b = 1;
				async function f() {
					await fetch();
				};
			}"
`;

exports[`noAsyncWithoutAwait fix Functions should fail when the function is not at the top level 1`] = `
"
			function a(){
				let b = 1;
				function f() {
					fetch();
				};
			}"
`;

exports[`noAsyncWithoutAwait fix return statements should fail when the async function has a return statement 1`] = `
"
			function a() {
				let b = 1;
				let a = () => {
				return 1;
				}
			}
		"
`;
